libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/navigation/NavigationStateViewModel.kt

```diff
@@ -0,0 +1,119 @@
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+/**
+ * Class that manages NavigationState as following:
+ *
+ * ```
+ *  [ FreeDrive ]━━( destination != null )━━━━━━>>[ RoutePreview ]<<━━━━━━━━━━━━┓
+ *       ^         ( & routes not empty  )               ┃                      ┃
+ *       ┃                                  (activeNavigationStarted=true)      ┃
+ *       ┃                                               ┃                      ┃
+ *       ┃                                               ┃         (activeNavigationStarted=false)
+ *       ┣━━━━( destination == null )━━━━┓               v                      ┃
+ *       ┗━━━━( routes empty        )━━━━╋━━━━━[ ActiveNavigation ]━━━━━━━━━━━━━┫
+ *                                       ┃               ┃                      ┃
+ *                                       ┃  (onFinalDestinationArrival())       ┃
+ *                                       ┃               ┃                      ┃
+ *                                       ┃               v                      ┃
+ *                                       ┗━━━━━━━━━━[ Arrival ]━━━━━━━━━━━━━━━━━┛
+ * ```
+ */
+internal class NavigationStateViewModel(
+    mutableState: MutableStateFlow<NavigationState>,
+    private val routesState: StateFlow<RoutesState>
+) : UIViewModel<NavigationState, NavigationStateAction>(mutableState) {
+
+    override fun process(
+        mapboxNavigation: MapboxNavigation,
+        state: NavigationState,
+        action: NavigationStateAction
+    ): NavigationState {
+        return when (action) {
+            is NavigationStateAction.Update -> action.state
+        }
+    }
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        mainJobControl.scope.launch {
+            routesState.collect {
+                val state = getNavigationState(
+                    it.destination,
+                    it.navigationStarted,
+                    mapboxNavigation.getRoutes()
+                )
+                updateState(state)
+            }
+        }
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowRoutesUpdated().collect {
+                val state = getNavigationState(
+                    routesState.value.destination,
+                    routesState.value.navigationStarted,
+                    it.routes
+                )
+                updateState(state)
+            }
+        }
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowOnFinalDestinationArrival().collect {
+                val inActiveNav =
+                    state.value == NavigationState.ActiveNavigation
+                val arrived = it.currentState == RouteProgressState.COMPLETE
+                if (inActiveNav && arrived) {
+                    updateState(NavigationState.Arrival)
+                }
+            }
+        }
```

Comment:
Convert each of these into `Action`s with parameters. You will be able to remove the circular dependency. Removing the dependency on the ViewModels